src/index.ts

```diff
@@ -70,67 +76,82 @@ export class Api {
+      this.clearTokens();
+      throw error;
+    }
   }
 
-  private refreshSessionJustBeforeExpiry() {
-    if (!this.tokenExpiration) return;
-
-    this.refreshTimeout = setTimeout(() => {
-      if (!this.refreshToken) {
-        this.initializeSession();
-        return;
-      }
-
-      const request = this.clientDict.oauth.tokens
-        .getToken(
-          {
-            // @ts-ignore
-            grant_type: 'refresh_token',
-            refresh_token: this.refreshToken,
-          },
-          {
-            type: ContentType.UrlEncoded,
-            headers: {
-              Authorization: `Basic ${this.basicAuthToken}`,
-            },
+  private async renewAuthToken() {
+    try {
+      const { data } = await this.clientDict.oauth.tokens.getToken(
+        {
+          // @ts-ignore
+          grant_type: 'refresh_token',
+          refresh_token: this.refreshToken,
+        },
+        {
+          type: ContentType.UrlEncoded,
+          headers: {
+            Authorization: `Basic ${this.basicAuthToken}`,
           },
-        )
-        .then(({ data }) => {
-          this.setTokenHeaders(data);
-          this.refreshSessionJustBeforeExpiry();
-        });
-
-      this.updatePrerequest(request);
-    }, this.tokenExpiration * 900);
+        },
+      );
+      this.setTokenHeaders(data);
+    } catch (error) {
+      console.error('Refreshing OHIP session token failed', error);
+      await this.requestNewAuthToken();
+    }
   }
 
   private setTokenHeaders(response: OAuth2TokenResponse) {
     // @ts-ignore
     this.refreshToken = response.refresh_token;
     this.token = response.access_token;
-    this.tokenExpiration = response.expires_in;
+    // response.expires_in is in number of seconds. Multiply it by 90% of 1000ms
+    this.tokenExpiration = Date.now() + (response.expires_in ?? 0) * 900;
+    this.retryLimit = 0;
   }
 
-  private updatePrerequest(request: Promise<any>) {
-    Object.entries(this.clientDict)
-      .filter(([id]) => id !== 'oauth')
-      .forEach(([_, client]) => {
-        // Every request waits for the prerequest to finish before starting the actual request
-        // This ensures that auth tokens are already available before the actual request is sent
-        client.prerequest = request;
-      });
+  private clearTokens() {
+    this.refreshToken = undefined;
+    this.token = undefined;
+    this.tokenExpiration = undefined;
+  }
+
+  private isAuthTokenExpired() {
+    return Date.now() > this.tokenExpiration;
   }
 
-  private handleClientRequest = (config: AxiosRequestConfig) => {
+  private handleClientRequest = async (config: AxiosRequestConfig) => {
+    if (!this.token) {
+      await this.requestNewAuthToken();
+    } else if (this.isAuthTokenExpired()) {
+      await this.renewAuthToken();
+    }
     config.headers['Authorization'] = `Bearer ${this.token}`;
 
     return config;
   };
+
+  private handleClientRequestError = async (error: any) => {
```

Comment:
do `error: unknown` instead of `any` then you can do something like:
```
var myError = new Error('foo');
myError instanceof Error // true
var myString = "Whatever";
myString instanceof Error // false
```

Reply:
You can also make the type of error in the `instaceof` more specific

Reply:
AxiosError doesn't have a class, so I can't use instanceof.
`unknown` warns an error if I access any of its properties.

What I did is create a function to determine if the error object is an AxiosError or not.
```typescript
function isAxiosError(error: any): error is AxiosError {
  return !!error.response;
}
```